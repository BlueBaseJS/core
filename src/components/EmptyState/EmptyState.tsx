import { BlueBase } from '../../BlueBase';
import { BlueBaseContext } from '../../Context';
import React from 'react';

export class EmptyState extends React.PureComponent {

	static contextType = BlueBaseContext;

	render() {

		// FIXME: remove typecasting, added because current react typings don't seem to support this.context
		const BB: BlueBase = (this as any).context;

		return (
			<BB.Components.ComponentState title="Empty area" description="No data found." />
		);
	}
}
