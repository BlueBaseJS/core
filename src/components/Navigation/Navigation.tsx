import { BlueBaseContext, NavigationContext, StubNavigationActionsObject } from '../../contexts';
import { NavigationProps, RouteConfig } from '@bluebase/components';

import { BlueBase } from '../../BlueBase';
import React from 'react';
import { resolveThunk } from '../../utils';

/**
 * ðŸ”€ Navigation
 *
 * This is a stub router. Intended to be replaced by an external router plugin.
 */
export class Navigation extends React.PureComponent<NavigationProps> {
	static contextType: React.Context<BlueBase> = BlueBaseContext;

	render() {
		const BB: BlueBase = this.context;

		const routes = resolveThunk<RouteConfig[]>(this.props.navigator.routes, BB) || [];

		const route = routes[0];

		if (!route) {
			return null;
		}

		const Screen =
			typeof route.screen === 'string' ? BB.Components.resolve(route.screen) : route.screen;

		let navigator = null;

		if (route.navigator) {
			navigator = <Navigation navigator={route.navigator} />;
		}

		let node = navigator;

		if (route.screen) {
			const Component = Screen!;
			node = <Component key={route.name}>{navigator}</Component>;
		}

		return (
			<NavigationContext.Provider value={StubNavigationActionsObject}>
				{node}
			</NavigationContext.Provider>
		);
	}
}
