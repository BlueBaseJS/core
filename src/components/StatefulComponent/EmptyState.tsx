import { BlueRainType, withBlueRain } from '../../index';
import React from 'react';

const EmptyState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Empty area"
			description="No data found."
		/>
	);
};

export default withBlueRain(EmptyState);
