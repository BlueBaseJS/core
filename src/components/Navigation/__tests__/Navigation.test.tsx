// tslint:disable: object-literal-sort-keys
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';
import React from 'react';
import { Text } from 'react-native';

import { BlueBaseApp } from '../../../';
import { EmptyState } from '../../EmptyState';
import { Navigation } from '../Navigation';

function SomeComponent() {
	return <Text>SomeComponent</Text>;
}

describe('Navigation', () => {
	test('should render a screen component', async () => {
		const navigator = {
			type: 'stack',
			routes: [
				{
					name: 'Home',
					path: '',
					screen: SomeComponent,
				},
			],
		};

		const component = mount(
			<BlueBaseApp>
				<Navigation navigator={navigator} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, SomeComponent);

		expect(component.exists('Navigation SomeComponent')).toBe(true);
	});

	test('should render a screen component from component registry', async () => {
		const navigator = {
			type: 'stack',
			routes: [
				{
					name: 'Home',
					path: '',
					screen: 'EmptyState',
				},
			],
		};

		const component = mount(
			<BlueBaseApp>
				<Navigation navigator={navigator} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, EmptyState);

		expect(component.exists('Navigation EmptyState')).toBe(true);
	});

	test('should render a screen component in a nested navigator', async () => {
		const navigator = {
			type: 'stack',
			routes: [
				{
					name: 'Home',
					path: '',
					navigator: {
						type: 'stack',
						routes: [
							{
								name: 'Nested',
								path: '',
								screen: SomeComponent,
							},
						],
					},
				},
			],
		};

		const component = mount(
			<BlueBaseApp>
				<Navigation navigator={navigator} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, SomeComponent);

		expect(component.exists('Navigation SomeComponent')).toBe(true);
	});

	test('should render null if theres no screen or navigator prop in route', async () => {
		const navigator = {
			type: 'stack',
			routes: [
				{
					name: 'Home',
					path: '',
				},
			],
		};

		const component = mount(
			<BlueBaseApp>
				<Navigation navigator={navigator} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, Navigation);

		expect(component.find('Navigation').children()).toHaveLength(0);
	});

	test('should render null if theres no route', async () => {
		const navigator = {
			type: 'stack',
		} as any;

		const component = mount(
			<BlueBaseApp>
				<Navigation navigator={navigator} />
			</BlueBaseApp>
		);

		// Wait for render
		await waitForElement(component as any, Navigation);

		expect(component.find('Navigation').children()).toHaveLength(0);
	});
});
