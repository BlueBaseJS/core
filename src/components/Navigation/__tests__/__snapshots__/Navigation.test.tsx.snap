// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation should render a screen component 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <Navigation
        navigator={
          Object {
            "routes": Array [
              Object {
                "name": "Home",
                "path": "",
                "screen": [Function],
              },
            ],
            "type": "stack",
          }
        }
      >
        <SomeComponent
          key="Home"
        >
          <Text>
            <Text>
              SomeComponent
            </Text>
          </Text>
        </SomeComponent>
      </Navigation>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`Navigation should render a screen component from component registry 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <Navigation
        navigator={
          Object {
            "routes": Array [
              Object {
                "name": "Home",
                "path": "",
                "screen": "EmptyState",
              },
            ],
            "type": "stack",
          }
        }
      >
        <ThemedComponent
          key="Home"
        >
          <EmptyState>
            <ComponentState
              description="Create new items to fill this area."
              title="Empty Area"
            >
              <ThemedComponent
                description="Create new items to fill this area."
                title="Empty Area"
              >
                <ThemedComponent
                  description="Create new items to fill this area."
                  title="Empty Area"
                >
                  <ComponentState
                    description="Create new items to fill this area."
                    styles={
                      Object {
                        "actionRoot": Object {
                          "marginTop": 8,
                        },
                        "description": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                        "image": Object {
                          "height": 250,
                          "width": 250,
                        },
                        "imageRoot": Object {
                          "marginBottom": 16,
                        },
                        "root": Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        },
                        "title": Object {
                          "marginBottom": 8,
                          "textAlign": "center",
                        },
                      }
                    }
                    title="Empty Area"
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "flex": 1,
                          "justifyContent": "center",
                        }
                      }
                    >
                      <ThemedComponent
                        style={
                          Object {
                            "alignItems": "center",
                            "flex": 1,
                            "justifyContent": "center",
                          }
                        }
                      >
                        <ThemedComponent
                          style={
                            Object {
                              "alignItems": "center",
                              "flex": 1,
                              "justifyContent": "center",
                            }
                          }
                        >
                          <ThemedComponent
                            style={
                              Object {
                                "alignItems": "center",
                                "flex": 1,
                                "justifyContent": "center",
                              }
                            }
                          >
                            <View
                              style={
                                Object {
                                  "alignItems": "center",
                                  "flex": 1,
                                  "justifyContent": "center",
                                }
                              }
                            >
                              <View
                                style={
                                  Object {
                                    "alignItems": "center",
                                    "flex": 1,
                                    "justifyContent": "center",
                                  }
                                }
                              >
                                <H6
                                  style={
                                    Object {
                                      "marginBottom": 8,
                                      "textAlign": "center",
                                    }
                                  }
                                >
                                  <ThemedComponent
                                    style={
                                      Object {
                                        "marginBottom": 8,
                                        "textAlign": "center",
                                      }
                                    }
                                  >
                                    <Component
                                      style={
                                        Object {
                                          "marginBottom": 8,
                                          "textAlign": "center",
                                        }
                                      }
                                      styles={
                                        Object {
                                          "root": Object {
                                            "color": "rgba(0, 0, 0, 0.87)",
                                            "fontFamily": undefined,
                                            "fontSize": 20,
                                            "fontWeight": "500",
                                          },
                                        }
                                      }
                                    >
                                      <Text
                                        style={
                                          Array [
                                            Object {
                                              "color": "rgba(0, 0, 0, 0.87)",
                                              "fontFamily": undefined,
                                              "fontSize": 20,
                                              "fontWeight": "500",
                                            },
                                            Object {
                                              "marginBottom": 8,
                                              "textAlign": "center",
                                            },
                                          ]
                                        }
                                      >
                                        <Text
                                          style={
                                            Array [
                                              Object {
                                                "color": "rgba(0, 0, 0, 0.87)",
                                                "fontFamily": undefined,
                                                "fontSize": 20,
                                                "fontWeight": "500",
                                              },
                                              Object {
                                                "marginBottom": 8,
                                                "textAlign": "center",
                                              },
                                            ]
                                          }
                                        >
                                          Empty Area
                                        </Text>
                                      </Text>
                                    </Component>
                                  </ThemedComponent>
                                </H6>
                                <Body2
                                  style={
                                    Object {
                                      "marginBottom": 8,
                                      "textAlign": "center",
                                    }
                                  }
                                >
                                  <ThemedComponent
                                    style={
                                      Object {
                                        "marginBottom": 8,
                                        "textAlign": "center",
                                      }
                                    }
                                  >
                                    <Component
                                      style={
                                        Object {
                                          "marginBottom": 8,
                                          "textAlign": "center",
                                        }
                                      }
                                      styles={
                                        Object {
                                          "root": Object {
                                            "color": "rgba(0, 0, 0, 0.87)",
                                            "fontFamily": undefined,
                                            "fontSize": 14,
                                            "fontWeight": "400",
                                          },
                                        }
                                      }
                                    >
                                      <Text
                                        style={
                                          Array [
                                            Object {
                                              "color": "rgba(0, 0, 0, 0.87)",
                                              "fontFamily": undefined,
                                              "fontSize": 14,
                                              "fontWeight": "400",
                                            },
                                            Object {
                                              "marginBottom": 8,
                                              "textAlign": "center",
                                            },
                                          ]
                                        }
                                      >
                                        <Text
                                          style={
                                            Array [
                                              Object {
                                                "color": "rgba(0, 0, 0, 0.87)",
                                                "fontFamily": undefined,
                                                "fontSize": 14,
                                                "fontWeight": "400",
                                              },
                                              Object {
                                                "marginBottom": 8,
                                                "textAlign": "center",
                                              },
                                            ]
                                          }
                                        >
                                          Create new items to fill this area.
                                        </Text>
                                      </Text>
                                    </Component>
                                  </ThemedComponent>
                                </Body2>
                              </View>
                            </View>
                          </ThemedComponent>
                        </ThemedComponent>
                      </ThemedComponent>
                    </View>
                  </ComponentState>
                </ThemedComponent>
              </ThemedComponent>
            </ComponentState>
          </EmptyState>
        </ThemedComponent>
      </Navigation>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`Navigation should render a screen component in a nested navigator 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <Navigation
        navigator={
          Object {
            "routes": Array [
              Object {
                "name": "Home",
                "navigator": Object {
                  "routes": Array [
                    Object {
                      "name": "Nested",
                      "path": "",
                      "screen": [Function],
                    },
                  ],
                  "type": "stack",
                },
                "path": "",
              },
            ],
            "type": "stack",
          }
        }
      >
        <Navigation
          navigator={
            Object {
              "routes": Array [
                Object {
                  "name": "Nested",
                  "path": "",
                  "screen": [Function],
                },
              ],
              "type": "stack",
            }
          }
        >
          <SomeComponent
            key="Nested"
          >
            <Text>
              <Text>
                SomeComponent
              </Text>
            </Text>
          </SomeComponent>
        </Navigation>
      </Navigation>
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`Navigation should render null if theres no route 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <Navigation
        navigator={
          Object {
            "type": "stack",
          }
        }
      />
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;

exports[`Navigation should render null if theres no screen or navigator prop in route 1`] = `
<BlueBaseApp>
  <BlueBaseRoot>
    <ThemeProvider>
      <Navigation
        navigator={
          Object {
            "routes": Array [
              Object {
                "name": "Home",
                "path": "",
              },
            ],
            "type": "stack",
          }
        }
      />
    </ThemeProvider>
  </BlueBaseRoot>
</BlueBaseApp>
`;
